### PR TITLE
merge vertices operator

### PR DESCRIPTION
**Pull request by nickeubank**
Here's a base attempt to merge vertices. First Julia PR, so pointers most welcome. 

Also, still working through test syntax...

### CODE REVIEW

File: src/operators.jl [hunk 1]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+Combines vertices specified in `vs` into single vertex whose
+index will be the lowest value in `vs`.
+
+All edges running to vertices in `vs` will connect to new
+merged vertex.
+
+Returns an Array{Int} with new vertex values for
+"""
+
+function merge_vertices!(g::Graph, vs::Array{Int})
```
Comment (sbromberger) [inline src/operators.jl]: `vs` should either be untyped or be a `Vector{T}` where `T <: Integer` since vertices don't have to be `Int`s.

File: src/operators.jl [hunk 2]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+    allunique(vs) || error("Vertices in vs must be unique")
```
Comment (sbromberger) [inline src/operators.jl]: This may be expensive.

File: src/operators.jl [hunk 3]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+    vertex_conversions = collect(1:nv(g))
```
Comment (sbromberger) [inline src/operators.jl]: Make this `collect(vertices(g))`

File: src/operators.jl [hunk 4]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+        for n in neighbors(g, v)
+            if v0 != n
+                add_edge!(g, v0, n)
+            end
+        end
+
+        # Now remove vertex
+        rem_vertex!(g, v)
+        vertex_conversions[v] = v0
+        vertex_conversions[(v+1):end] = vertex_conversions[(v+1):end] - 1
```
Comment (jpfairbanks) [inline src/operators.jl]: I think this might cause `O(n*length(vs))` performance because you are updating the `vertex_conversions` every time around the loop `length(vs)` times. Is there a better way to store this information? I think you can compute the final value of `vertex_conversions[i]` based on the number of elements in `vs` that is less that `i`.

Comment (sbromberger) [inline src/operators.jl]: This is expensive, and I'm not sure what its purpose is.

File: src/operators.jl [hunk 5]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+function merge_vertices!(g::Graph, vs::Array{Int})
+
+    allunique(vs) || error("Vertices in vs must be unique")
+
+    vertex_conversions = collect(1:nv(g))
+
+    # Use lowest value as new vertex id.
+    v0 = minimum(vs)
+    filter!(x->x != v0, vs)
```
Comment (jpfairbanks) [inline src/operators.jl]: Does this require length(vs) operations to remove the one element?
When you do `allunique` it uses a Set to determine the unique elements.
You could instead use `sort!(vs)` and then use `vs = vs[2:end]`

**Comment by jpfairbanks** [comment]
Seth and I had some overlapping thoughts :)

File: src/operators.jl [hunk 6]
```diff
@@ -442,3 +442,43 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
 with respect to `v` (i.e. `:in` or `:out`).
 """
 egonet(g::AbstractGraph, v::Integer, d::Integer; dir=:out) = g[neighborhood(g, v, d, dir=dir)]
+
+
+"""
+    merge_vertices!(g::Graph, vs::Array{Int})
```
Comment (nickeubank) [inline src/operators.jl]: tweaked.

File: src/operators.jl (hunk 4 above)
Comment (nickeubank) [inline src/operators.jl]: Thought it was downshifting indices. Understand now.

Comment (nickeubank) [inline src/operators.jl]: Fixed that -- left that in under impression that removing a vertex lead to a down-shifting of indices rather than the swap that apparently takes place.

Comment (jpfairbanks) [inline src/operators.jl]: Yeah the shifting is too expensive for this reason. That is why we swap

File: src/operators.jl (hunk 3 above)
Comment (jpfairbanks) [inline src/operators.jl]: This is to support graphs that are not numbered starting at 1.

File: src/operators.jl (hunk 1 above)
Comment (nickeubank) [inline src/operators.jl]: Ah, ok. Why no AbstractInt type to make that easier?

**Comment by nickeubank** [comment]
@sbromberger @jpfairbanks updated and squashed.

File: src/operators.jl (hunk 1 above)
Comment (sbromberger) [inline src/operators.jl]: `Integer` is abstract.

File: test/operators.jl [hunk 7]
```diff
@@ -37,6 +37,16 @@
...
+
+        # Check merge_vertices function.
+        h = Graph{T}(7)
+        add_edge!(h, 2, 5)
+        add_edge!(h, 3, 6)
+        add_edge!(h, 1, 7)
+        new_map = merge_vertices!(h, [2, 3, 7, 3, 3, 2])
+        @test new_map == [1, 2, 2, 4, 5, 6, 2]
+        @test neighbors(h, 2) == [1, 3, 5]
+
```
Comment (sbromberger) [inline test/operators.jl]: you'll want to check DiGraphs as well.

**Comment by sbromberger** [comment]
You'll need to export merge_vertices! (check `src/LightGraphs.jl`) in order for this to be used.

**Comment by nickeubank** [comment]
@sbromberger Added DiGraph support, expanded tests. If you can think of more tests, wouldn't mind adding -- the whole "swap with top vertex and then drop top vertex" is very tricky. Pretty sure I got it pinned down, but it's slippery enough I'm amenable to more tests.

**Comment by nickeubank** [comment]
Nvm -- realized a corner case that will cause problems. Will update.

**Comment by jpfairbanks** [comment]
I am concerned about using `rem_vertex!` to do these merges. It leads to a vertex labeling in the final graph that is dependent on the order in which you process `vs`.

When you remove vertex `i` you make the current vertex `nv` into vertex `i` and then move on to the next `i` in `vs`. If we choose the convention of processing `vs` in sorted order, then the final ordering is deterministic. But if we process `vs` in arbitrary order, as would be necessary for parallel processing, then the final vertex labels are not deterministic.

If we were to use this labeling for the final output, then we should provide a deterministic ordering and a function to calculate it ahead of time. 

I believe that function is 
```julia
function swaplabel(vs::AbstractArray, n::Integer, v::Integer)
   if v in vs
      return minimum(vs)
   else if v == n-i
      return vs[i+1] # +1 because we don't remove vs[1]
   else
      return v
   end
end
```

We could choose that ordering, but it seems overly complicated compared to the shifting down rule.

```julia
function shiftlabel(vs, n::Integer, v::Integer)
     x = zeros(Int, n)
     x[vs] = 1
     return collect(1:n).-cumsum(x)
end

I am ok with choosing either ordering as the official definition. But this ordering is part of the interface of the `merge_vertices!(g)` function and changing it will be considered "breaking the interface." We need to pick a definition and stick with it, and provide a function to compute the ordering so that we can enforce stability of the interface when changing the implementation.

**Comment by nickeubank** [comment]
@jpfairbanks At the moment, I sort `vs` before using it, so it is still deterministic regardless of order of `vs`. 

I'm amenable to a shift-strategy (the book-keeping of the order jumping has definitely been trickiest part), but I'm also generally of the view a library shouldn't allow a proliferation of tools. If removing vertices has a set function (in this case, `rem_vertex`, seems like we should always use it, no?

**Comment by nickeubank** [comment]
@sbromberger @jpfairbanks fixed that bug I'd mentioned.

**Comment by jpfairbanks** [comment]
We can use `rem_vertex!` but it isn't really imperative. The concept of merging vertices is big enough that it can get its own functions.

The reason for using the swapping semantics for `rem_vertex!` is purely for performance. We can use shifting semantics for merging, which will be more common in the literature for community detection and graph partitioning anyway.

The reason I came to think about this was I started to implement a non-modifying version of merge_vertices. That is, a function that made a new graph where the vertices had been merged. I figured that it was simpler to implement correctly and could be used to verify your in-place (modifying) implementation.

When I did that I realized that swapping semantics are unnatural for a bulk operation where you need to remove many vertices at once. Shifting semantics are more natural in this case.

**Comment by jpfairbanks** [comment]
Do you feel like you have a good handle on the bookkeeping for swapping semantics? Can you do that bookkeeping ahead of time so that you can say what the labels will be before you modify the graph?

**Comment by nickeubank** [comment]
Not realistically, no -- I could probably sit down and wrestle through it, but no user ever would -- you'd be reliant on the output vertex to give you your new vertex mappings.

**Comment by jpfairbanks** [comment]
If we can't commit to having the same ordering for swapping semantics even if we change the implementation, then I would rather use the shifting semantics which are easy to calculate and parallelize.

**Comment by jpfairbanks** [comment]
hmm, that leaves us with a problem where we need to directly manipulate the adjacency list structure which adds to the complexity of the operation. Not all graph datastructures support rem_vertex with the same ordering semantics being the most efficient.

**Comment by nickeubank** [comment]
I guess the performance issue still seems salient for me, especially if a core use of this is for community detection, for which performance is often a huge constraint. 

More broadly, seems like one important question here is the philosophy of managing labels in the library. My sense from @sbromberger was that the library had taken a "we'll do what we want, and return a vector with new labels" strategy. I'm not sure I love that (I work with vertex attributes, so it's bookkeeping I'll always have to do in my own work), but if the library is committed to that, then I don't know that we need to also commit to stable ordering of outputs. 

Alternatively, I could wrestle through making the new labels in a stand-alone way -- it _is_ deterministic, so I could do it, and we could use that if we change the implementation.

**Comment by nickeubank** [comment]
So what is the general philosophy on vertex book-keeping for the library?

**Comment by jpfairbanks** [comment]
Yes I agree that returning the new labels is the only way to avoid imposing a standard of how the ordering is computed.

For community detection is it feasible that you would create a new instance of the graph with the vertices merged rather than modifying the original graph? 

I can see a lot of use cases where I would want to make a copy of the graph before doing that anyway because I will need the original graph for some other operation later. For example I am going to do community detection and then for each community solve a ranking problem on the graph restricted to that community. In this example, I would need to do `labels = community_detection!(copy(g))` so that I can work with `g` to do the ranking.

**Comment by jpfairbanks** [comment]
My philosophy is to minimize the amount of vertex bookkeeping because it introduces opportunities for error. And to store the mapping from old labels to new labels in an array inside some struct whereever necessary.

**Comment by sbromberger** [comment]
> My sense from @sbromberger was that the library had taken a "we'll do what we want, and return a vector with new labels" strategy.

Well, it's more of a strategy of "let's do the thing that's most performant", and the swap-n-pop method is the fastest for removal of vertices so far (we didn't do it this way in the beginning). I'm open to alternatives as long as we keep the 3 design goals in mind.

**Comment by nickeubank** [comment]
> For community detection is it feasible that you would create a new instance of the graph with the vertices merged rather than modifying the original graph?

Maybe I'm showing my Python background, but seems like the easiest way to do this is just open the function with a `copy()` command of the input graph (and drop the bang from the name). My intuition is that doing a direct copy and then making a few adjustments would be faster than writing out a full new graph edge by edge. But maybe that's bad intuition for Julia?

**Comment by sbromberger** [comment]
> My intuition is that doing a direct copy and then making a few adjustments would be faster than writing out a full new graph edge by edge. 

For a copy, it might be (depends on how many changes you want to make), but this wouldn't scale fore very large graphs as you'd have two structures in memory.

**Comment by nickeubank** [comment]
The scale problem seems inescapable if we want it to not modify the original graph. I suppose with LOTS of merged vertices building the new graph from scratch is better (since the new graph will be substantially smaller), but for any moderate numbers of merges, you'll end up without roughly two graphs either way.

**Comment by jpfairbanks** [comment]
Right, a copy in this case takes about as long as creating a new graph and filling it in. There is no "copy on write" magic to help us here.

It is better to have the bang version and a nonbang version so that the advanced user can avoid the wasted overhead of the nonbang version.

The only way to get around 2x memory for community detection where you want to keep the original graph is to do the IO again after you finish building your community hierarchy.

**Comment by jpfairbanks** [comment]
If you aren't doing large merges at once then you are going to have bad performance anyway. You need to be doing something like cutting the size of the graph in half at each step in order to achieve efficient algorithms.

**Comment by nickeubank** [comment]
OK, so sounds like we'd like to have both a bang version, and a non-bang version that builds from the bottom up. 

But we're still discussion importance of predictable vertex ordering?

To get predictable, seems like we have three options:

- Build new from scratch (which doesn't work for the bang version)
- Use `rem_vertex!` then try and re-structure to give shift-ordering
- Create a new shift-ordering-based version of `rem_vertex` and use that. 
- Accept the order won't be shift-ordered
    - accept and leave it as it is
    - write an algorithm to predict the new order and put that at the head of the function and force any future implementation to stick to it (though if, as you say, `rem_vertex!` does not promise a specific order of output labels, then I'd also have to add checks to make sure that order is always preserved)


I'm nervous about the last one given the performance implications

**Comment by sbromberger** [comment]
ok, wait. Why can't we simply do this?

Given a graph `g` and set of vertices `vs` to merge:

```
  component = minimum(vs)
  for i in 1:nv(g)
    nbrs = Set{eltype(g)}()
    for j in g.fadjlist[i]
    if j in vs
      push!(nbrs, component)
    else
      push!(nbrs, j)
    end
    g.fadjlist[i] = sort(collect(nbrs))
end
```

**Comment by nickeubank** [comment]
@sbromberger But that never deletes the extra vertices, does it? They're there, just without edges?

**Comment by sbromberger** [comment]
@nickeubank (check my edits :) )

**Comment by nickeubank** [comment]
@sbromberger :) thanks. 

That seems roughly the same as my current implementation (albeit more elegant due to use of fadjlist). Which concern does that address? Seems like we still have the re-ordering problem @jpfairbanks brought up to start this discussion when we drop the vertices...

**Comment by jpfairbanks** [comment]
@sbromberger's algorithm needs to decrement j to account for the shifting semantics.

```julia
 component = minimum(vs)
 shift = begin 
     x = zeros(Int, n)
     x[vs] = 1
     collect(1:n).-cumsum(x)
  end
  for i in 1:nv(g)
    nbrs = Set{eltype(g)}()
    for j in g.fadjlist[i]
    if j in vs
      push!(nbrs, component)
    else
      push!(nbrs, j-shift[j])
    end
    g.fadjlist[i] = sort(collect(nbrs))
end
g.ne = sum((x->degree(g,x).(vertices(g))/2
```

But that would work. It is tied to the implementation of SimpleGraph as a fadjlist. but that works for now. We can decide to enforce shifting semantics and then come up with an efficient implementation for other datastructures later.

If we don't want to enforce an ordering on `rem_vertex!` then we shouldn't enforce and ordering on `merge_vertices` and so we have to just roll with an approach of returning the mapping from old to new labels for every function that changes the ordering.

Also the set is inefficient, because you can't reuse the memory.

**Comment by sbromberger** [comment]
@jpfairbanks we can possibly rewrite using the API (`*_neighbors`) and minimize the datastructure-specific access.

**Comment by jpfairbanks** [comment]
Well the real data structure specific thing is that we are able to set the neighbors of a vertex in bulk.
That could be done with `rem_edge!(g, v,u)` a bunch of times and `add_edge(g,v,u)` the new edges. So instead of removing vertices we are "transferring" the edges.

Maybe the right answer is to have one implementation for `AbstractGraph` that is using `add/rem_edge` and then another implementation for `SimpleGraph` that uses direct access to the `fadjlist` to get better performance.

**Comment by sbromberger** [comment]
> Maybe the right answer is to have one implementation for AbstractGraph that is using add/rem_edge and then another implementation for SimpleGraph that uses direct access to the fadjlist to get better performance.

To what extent do we need to ensure that the vertex mappings are consistent between the "generic" version and any optimized ones?

**Comment by jpfairbanks** [comment]
I think they need to be consistent

**Comment by sbromberger** [comment]
Then we should rethink `swap-n-pop`. Shuffle-down is the most intuitive, but I don't know of a good way to do this efficiently.

**Comment by jpfairbanks** [comment]
I think swap-n-pop makes sense for removing individual vertices, but for more complex operations where you need to remove large numbers of vertices it makes sense to shift them down.

Or really I think about the situation as creating a different graph where you have applied some transformation rather than mutating an existing graph.

**Comment by sbromberger** [comment]
I keep coming back to this:
```
rem_vertices!(g::T, vs) where T <: AbstractGraph = T(adjacency_matrix(g)[setdiff(1:end, vs), setdiff(1:end, vs)])
```

It's not the most efficient, but it's certainly the easiest to grok.

```
julia> g = Graph(1_000_000, 5_000_000)
{1000000, 5000000} undirected simple Int64 graph

julia> v2r = unique(rand(1:1_000_000, 300_000));

julia> length(v2r)
259150

julia> @time h = rem_vertices!(g, v2r)
  6.201711 seconds (11.84 M allocations: 1.105 GiB, 44.78% gc time)
{740850, 2742213} undirected simple Int64 graph

**Comment by nickeubank** [comment]
In a python library, the solution would be write the function with an option for whether to do swap & pop or down-shift, set the default to the more intuitive one (down-shift), but make the fast one (swap & pop) available. Is that Julia-style consistent?

**Comment by sbromberger** [comment]
> In a python library, the solution would be write the rem_vertex! function with an option for whether to do swap & pop or down-shift, set the default to the more intuitive one (down-shift), but make the fast one (swap & pop) available. Is that Julia-style consistent?

It might be but note that `rem_vertex!` (swap-n-pop) is mutating, while `rem_vertices` (downshift) is not. There's no really efficient way to do `rem_vertices!` for a general graph, and particularly with an adjacency list representation. (It's different with SimpleWeightedGraphs since the backend is a sparse matrix.)

**Comment by nickeubank** [comment]
I will confess to have gotten a little turned around at this point between discussions of library design goals and what needs to happen in this PR. Could one of you summarize where we stand on the design of this function?

**Comment by jpfairbanks** [comment]
For this PR we came to a consensus that down-shift is the better ordering and so we need an implementation of `merge_vertices!` that produces a graph with down-shifting semantics. With tests.


From there we can move on to the next PR which will have an agglomerative community detection algorithm.

**Comment by jpfairbanks** [comment]
See commit: d4bdf4a for my out of place implementation. You can your in place version against the out of place version. If they agree then we can merge them both.

File: src/operators.jl [hunk 8]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+Supports Graphs and DiGraphs.
+
+All edges running to vertices in `vs` connect to new
+merged vertex.
+
+Return a vector with new vertex values index by original vertex
+indices.
+"""
+
+function merge_vertices!(g::AbstractGraph, vs::Vector{T} where T <: Integer)
```
Comment (sbromberger) [inline src/operators.jl]: This will only work with `SimpleGraph`.

File: src/operators.jl [hunk 9]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+    vs = sort!(unique(vs))
+    merged_vertex = vs[1]
+    vs = vs[2:end]
```
Comment (sbromberger) [inline src/operators.jl]: the above lines can be replaced with
```
merged_vertex = shift!(vs)
```

File: src/operators.jl [hunk 10]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+    x = zeros(Int, nv(g))
+    x[vs] = 1
+    new_vertex_ids = collect(1:nv(g)) .- cumsum(x)
+    new_vertex_ids[vs] = merged_vertex
+
+    for i in 1:nv(g)
```
Comment (sbromberger) [inline src/operators.jl]: `for i in vertices(g)`

File: src/operators.jl [hunk 11]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+        # Adjust connections to merged vertices
+        if (i != merged_vertex) & !(i in vs)
+            nbrs_to_rewire = Set{eltype(g)}()
+            for j in g.fadjlist[i]
+               if j in vs
```
Comment (sbromberger) [inline src/operators.jl]: consider `for j in filter(x->x in vs, g.fadjlist[i])`

File: src/operators.jl [hunk 12]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+                 push!(nbrs_to_rewire, new_vertex_ids[j])
+               end
+            end
+            g.fadjlist[new_vertex_ids[i]] = sort(collect(nbrs_to_rewire))
+
+        # Collect connections to new merged vertex
+        else
+            nbrs_to_merge = Set{eltype(g)}()
+            for element in g.fadjlist[i]
+                if !(element in vs)
```
Comment (sbromberger) [inline src/operators.jl]: see `filter` example above.

File: src/operators.jl [hunk 13]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+                    push!(nbrs_to_merge, new_vertex_ids[element])
+                end
+            end
+
+            for j in vs
+                for e in g.fadjlist[j]
+                    if e != merged_vertex
```
Comment (sbromberger) [inline src/operators.jl]: I think you might be able to combine the two `for` statements on a single line:
`for j in vs, e in g.fadjjlist[j]`

File: src/operators.jl [hunk 14]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+            g.fadjlist[i] = sort(collect(nbrs_to_merge))
+        end
+    end
+
+    # Drop excess vertices
+    g.fadjlist = g.fadjlist[1: (end-length(vs))]
+
+    # Correct edge counts
+    total_edges = 0
+    for i in 1:nv(g)
```
Comment (sbromberger) [inline src/operators.jl]: `for i in vertices(g)`

File: src/operators.jl [hunk 15]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+        total_edges = total_edges + degree(g,i)
+    end
+
+    if isa(g, Graph)
+        g.ne = total_edges / 2
```
Comment (sbromberger) [inline src/operators.jl]: `if !is_directed(g)`

File: src/operators.jl [hunk 16]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+    elseif isa(g, DiGraph)
+        g.ne = total_edges
+    else:
```
Comment (sbromberger) [inline src/operators.jl]: Don't need this.

File: README.md [hunk 17]
```diff
@@ -108,6 +108,7 @@ These functions are defined as the public contract of the `LightGraphs.AbstractG
 - `rem_edge!`
 - `add_vertex!`, `add_vertices!`
 - `rem_vertex!`
+- `merge_vertices!H`
```
Comment (sbromberger) [inline README.md]: This is part of `operators`, not the core API.

File: src/operators.jl [hunk 18]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
 egonet(g::AbstractGraph, v::Integer, d::Integer; dir=:out) = g[neighborhood(g, v, d, dir=dir)]
+
+
+"""
+    merge_vertices!(g, vs)
+
+Combine vertices specified in `vs` into single vertex whose
+index will be the lowest value in `vs`.
+
+Supports Graphs and DiGraphs.
```
Comment (sbromberger) [inline src/operators.jl]: Validate this.

File: src/operators.jl [hunk 19]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+
+All edges running to vertices in `vs` connect to new
+merged vertex.
+
+Return a vector with new vertex values index by original vertex
```
Comment (sbromberger) [inline src/operators.jl]: `indexed`?

File: src/operators.jl [hunk 20]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+indices.
+"""
+
+function merge_vertices!(g::AbstractGraph, vs::Vector{T} where T <: Integer)
+    vs = sort!(unique(vs))
+    merged_vertex = vs[1]
+    vs = vs[2:end]
+
+    x = zeros(Int, nv(g))
```
Comment (sbromberger) [inline src/operators.jl]: Why `Int`? Won't this be the same as `eltype(g)`?

File: src/operators.jl [hunk 21]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+    x[vs] = 1
+    new_vertex_ids = collect(1:nv(g)) .- cumsum(x)
+    new_vertex_ids[vs] = merged_vertex
+
+    for i in 1:nv(g)
+
+        # Adjust connections to merged vertices
+        if (i != merged_vertex) & !(i in vs)
```
Comment (sbromberger) [inline src/operators.jl]: This will be O(n^2) unless you make a set out of `vs`.

File: src/operators.jl [hunk 22]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+            nbrs_to_rewire = Set{eltype(g)}()
+            for j in g.fadjlist[i]
```
Comment (sbromberger) [inline src/operators.jl]: `for j in out_neighbors(g, i)`

File: src/operators.jl [hunk 23]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+               else
+                 push!(nbrs_to_rewire, new_vertex_ids[j])
+               end
+            end
+            g.fadjlist[new_vertex_ids[i]] = sort(collect(nbrs_to_rewire))
+
+        # Collect connections to new merged vertex
+        else
+            nbrs_to_merge = Set{eltype(g)}()
+            for element in g.fadjlist[i]
```
Comment (sbromberger) [inline src/operators.jl]: `for element in out_neighbors(g, i)`

File: src/operators.jl [hunk 24]
```diff
@@ -442,3 +442,83 @@ This is equivalent to [`induced_subgraph`](@ref)`(g, neighborhood(g, v, d, dir=d
...
+    end
+
+    # Drop excess vertices
+    g.fadjlist = g.fadjlist[1: (end-length(vs))]
+
+    # Correct edge counts
+    total_edges = 0
+    for i in 1:nv(g)
+        total_edges = total_edges + degree(g,i)
+    end
```
Comment (sbromberger) [inline src/operators.jl]: `total_edges += sum(degree(g))`?

File: src/operators.jl (hunk 11 above)
Comment (sbromberger) [inline src/operators.jl]: Also - you might want to time `intersect(out_neighbors(g, i), vs)` as well. This might not be too performant.

File: src/operators.jl (hunk 8 above)
Comment (nickeubank) [inline src/operators.jl]: I get an error with `SimpleGraph` so putting in `Graph`

File: src/operators.jl (hunk 11 above)
Comment (nickeubank) [inline src/operators.jl]: I don't think that works here since it's a full `if`/`else` for `j in vs`, not just a filter.

**Comment by nickeubank** [comment]
@jpfairbanks @sbromberger OK, I think that covers all comments.

Since squashing crushed previous implementation built on swap-and-go, I'm gonna leave it here in case its ever of future use:

    function merge_vertices!(g::AbstractGraph, vs::Vector{T} where T <: Integer)
	 
	    vertex_conversions = collect(vertices(g))

	    # Use lowest value as new vertex id.
	    sort!(vs)
	    v0 = vs[1]

	    for v in vs[2:end]
	        new_v = vertex_conversions[v]

	        if new_v != v0

	            # Copy over all connections without adding self-loops
	            if isa(g, Graph)
	                for n in neighbors(g, new_v)
	                    if v0 != n
	                        add_edge!(g, v0, n)
	                    end
	                end

	            elseif isa(g, DiGraph)
	                # Copy over all connections without adding self-loops
	                for n in out_neighbors(g, new_v)
	                    if v0 != n
	                        add_edge!(g, v0, n)
	                    end
	                end
	                for n in in_neighbors(g, new_v)
	                    if v0 != n
	                        add_edge!(g, n, v0)
	                    end
	                end

	            else
	                error("merge_vertices only supports Graphs and DiGraphs")
	            end

	            # Now remove vertex
	            max_vertex = find(x -> x == nv(g), vertex_conversions)
	            vertex_conversions[max_vertex] = new_v
	            rem_vertex!(g, new_v)
	            vertex_conversions[v] = v0
	        end
	    end

	    return vertex_conversions
	end

**Comment by nickeubank** [comment]
closed by #685
